src/journal.rs

```diff
@@ -0,0 +1,181 @@
+
+#[derive(Debug, Clone, PartialEq, PartialOrd, Default)]
+struct JournalEntry {
+  ops: Vec<JournalOp>,
+  scope: usize,
+  scoped: bool,
+}
+
+impl JournalEntry {
+  fn new(ops: Vec<JournalOp>, scope: usize, scoped: bool) -> Self {
+    Self { ops, scope, scoped }
+  }
+}
+
+#[derive(Debug, Clone, PartialEq, PartialOrd)]
+pub enum JournalOp {
+  Call(Expr),
+  FnCall(Expr),
+  Push(Expr),
+  Pop(Expr),
+
+  FnStart(bool),
+  FnEnd,
+
+  Commit,
+}
+
+#[derive(Debug, Clone, PartialEq, PartialOrd, Default)]
+pub struct Journal {
+  ops: Vec<JournalOp>,
+  current: Vec<JournalOp>,
+  commits: Vec<usize>,
+}
+
+impl fmt::Display for Journal {
+  fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+    let entries = self.entries(20);
+    if !entries.is_empty() {
+      write!(f, "\n\nStack History (most recent first):\n")?;
+    }
+
+    let len = entries.len();
+    for (i, entry) in entries.iter().enumerate() {
+      let mut line = String::new();
+      let mut should_print = false;
+      for op in entry.ops.iter() {
+        if !line.is_empty() {
+          line.push(' ');
+        }
+
+        match op {
+          JournalOp::Call(call) => {
+            line.push_str(&format!("{}", color::Fg(color::White)));
+            line.push_str(&format!("{}", call));
+
+            line.push_str(&format!("{}", color::Fg(color::White)));
+            line.push_str(" |");
+
+            should_print = true;
+          }
+          JournalOp::FnCall(fn_call) => {
+            line.push_str(&format!("{}", color::Fg(color::Yellow)));
+            line.push_str(&format!("{}", fn_call));
+
+            line.push_str(&format!("{}", color::Fg(color::White)));
+            line.push_str(" |");
+
+            should_print = true;
+          }
+          JournalOp::Push(push) => {
+            line.push_str(&format!("{}", color::Fg(color::Green)));
+            line.push_str(&format!("{}", push));
+
+            should_print = true;
+          }
+          JournalOp::Pop(pop) => {
+            line.push_str(&format!("{}", color::Fg(color::Red)));
+            line.push_str(&format!("{}", pop));
+
+            should_print = true;
+          }
+          _ => {}
+        }
+      }
+
+      if should_print {
+        line.push_str(&format!("{}", color::Fg(color::Reset)));
+
+        let bullet_symbol = match entry.scoped {
+          true => format!("{}*", "  ".repeat(entry.scope)),
+          false => {
+            format!("{}!", "  ".repeat(entry.scope))
+          }
+        };
+        write!(f, " {} ", bullet_symbol)?;
+
+        if i != len - 1 {
+          writeln!(f, "{}", line)?;
+        } else {
+          write!(f, "{}", line)?;
+        }
+      }
+    }
+
+    Ok(())
+  }
+}
+
+impl Journal {
+  pub fn new() -> Self {
+    Self::default()
+  }
+
+  pub fn op(&mut self, op: JournalOp) {
+    if matches!(op, JournalOp::Commit) {
+      return self.commit();
+    }
+
+    self.current.push(op);
+  }
+
+  pub fn commit(&mut self) {
```

Comment:
This should be a ring buffer where old commits are removed once the max is reached

Reply:
TODO for later